### PR TITLE
Labeled data's instance wise contrastive loss

### PR DESCRIPTION
**Pull request by adv010**
Hi Farzad, raising a Pull request for the labeled instancewise loss. 

The repo used for this code was - https://github.com/HobbitLong/SupContrast/blob/master/losses.py
Link to the paper - https://arxiv.org/pdf/2004.11362.pdf

Note - I raised the PR with the branch containing non-linted code, as it was messing up the git diff and making the changes hard to read on PR. 

 Instead, I've created another branch  'linted_lb_inst_loss'' in my [repo](https://github.com/adv010/OpenPCDet/tree/linted_lb_inst_loss) .  Please refer to this branch for the PEP8 formatted code.

### CODE REVIEW

File: pcdet/models/detectors/pv_rcnn_ssl.py [hunk 1]
```diff
@@ -295,12 +295,10 @@ def _forward_training(self, batch_dict):
             if self.model_cfg.ROI_HEAD.INSTANCE_CONTRASTIVE_LOSS_MODEL=='Teacher':
                 lbl_inst_cont_loss = self._get_instance_contrastive_loss(batch_dict_sa,batch_dict_ema,lbl_inds,ulb_inds)
             elif self.model_cfg.ROI_HEAD.INSTANCE_CONTRASTIVE_LOSS_MODEL =='Student' :
-                lbl_inst_cont_loss = self._get_instance_contrastive_loss(batch_dict,batch_dict_wa,lbl_inds,ulb_inds)
+                lbl_inst_cont_loss, tb_dict = self._get_instance_contrastive_loss(tb_dict,batch_dict,batch_dict_wa,lbl_inds,ulb_inds)
```
Comment (frezaeix) [inline pcdet/models/detectors/pv_rcnn_ssl.py:298]: what is tb_dict?

Comment (frezaeix) [reply pcdet/models/detectors/pv_rcnn_ssl.py:298]: @adv010

File: pcdet/models/detectors/pv_rcnn_ssl.py [hunk 2]
```diff
@@ -342,15 +340,17 @@ def _get_proto_contrastive_loss(self, batch_dict, bank, ulb_inds):
 
 
     def _align_instance_pairs(self, batch_dict,batch_dict_pair,indices):
-        shared_ft_sa = batch_dict['shared_features_gt'].view(batch_dict['batch_size'],-1,256)[indices]
-        shared_ft_wa = batch_dict_pair['shared_features_gt'].view(batch_dict['batch_size'],-1,256)[indices]
+        
+        embed_size = 256 if not self.model_cfg['ROI_HEAD']['PROJECTOR'] else 256 # if possible, 128
```
Comment (frezaeix) [inline pcdet/models/detectors/pv_rcnn_ssl.py:344]: if and else produce the same thing! @adv010

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 3]
```diff
@@ -195,8 +195,7 @@ def forward(self, batch_dict, test_only=False,use_gtboxes=False):
                 if self.model_cfg['NORMALIZATION']:
                     # pooled_features dim : [GT_boxes,27648]
                     pooled_features = F.normalize(pooled_features, dim = -1)
-                if self.model_cfg['PROJECTOR']:
```
Comment (frezaeix) [inline pcdet/models/roi_heads/pvrcnn_head.py]: Can I ask what is projector config that you are ignoring it here? @adv010

File: pcdet/models/roi_heads/pvrcnn_head.py [hunk 4]
```diff
@@ -186,19 +189,15 @@ def forward(self, batch_dict, test_only=False,use_gtboxes=False):
...
             if self.model_cfg.ENABLE_INSTANCE_SUP_LOSS==True and start_epoch<=batch_dict['cur_epoch']<stop_epoch: # normalize embedding and produce projected representation only when instance_sup_loss true.
-
                 if self.model_cfg['NORMALIZATION']:
-                    pooled_features_copy = F.normalize(pooled_features, dim = -1)
-                    if self.model_cfg['PROJECTOR']:
-                        shared_features_gt = self.projector_fc_layer(pooled_features_copy.view(batch_size_rcnn, -1, 1))
-                else:
-                    shared_features_gt = self.projector_fc_layer(pooled_features.view(batch_size_rcnn, -1, 1))
-                batch_dict['shared_features_gt'] = shared_features_gt
+                    # pooled_features dim : [GT_boxes,27648]
```
Comment (frezaeix) [inline pcdet/models/roi_heads/pvrcnn_head.py:197]: Can you explain this part here because in the next commit you also changed it? @advait

File: pcdet/models/detectors/pv_rcnn_ssl.py (hunk 1 above)
Comment (adv010) [reply pcdet/models/detectors/pv_rcnn_ssl.py:298]: @frezaeix That's the dictionary being updated for tensorboard logging

File: pcdet/models/detectors/pv_rcnn_ssl.py (hunk 2 above)
Comment (adv010) [reply pcdet/models/detectors/pv_rcnn_ssl.py:344]: Yes, that could be ignored for now :) 
Currently, the setting I use is -  Both  shared_fc layer and projected_fc layer  having : 2 hidden layers with output neurons 256,256
In case we may want to try later for a lower-dimensional embedding with projected_fc layer as 256,128, then this embed_size would change to 128.

File: pcdet/models/roi_heads/pvrcnn_head.py (hunk 3 above)
Comment (adv010) [reply pcdet/models/roi_heads/pvrcnn_head.py]: @frezaeix During initial phases of implementing this loss,  I thought of projection as a good-to-have. However, in almost all contrastive literature, projection is a necessary step to have. Hence, I decided to remove this condition and make it part of the main pipeline.

The normalization is a step that I've picked up when reading the Supervised Contrastive loss paper, where they normalize features at 2 steps :
1.  The features from last layer of the backbone (ResNet)
Since this is new functionality we'll be introducing to the PVRCNN backbone, I chose to make it as an optional setting with the normalize flag.
2. Normalizing the projected embedding 
This is handled  within the contrastive_loss function itself

MicrosoftTeams-image (2)
This image shows the mental model of the current pipeline
